### PR TITLE
Do not crash if critical file is not found

### LINKED ISSUE #486: Could not read temperat.pal
I just cloned the master branch and compiled it. Then I placed the `vanillatd` binary into the CnCNet directory and launched it, which results in a Segmentation Fault with the following stacktrace
```
(gdb) backtrace
#0  0x00007ffff79fc679 in clearerr (fp=0x0) at clearerr.c:25
#1  0x0000555555680f67 in RawFileClass::Raw_Seek (this=0x7fffffffd790, pos=0, 
    dir=1) at /home/giulianob/bagunca/Vanilla-Conquer/common/rawfile.cpp:886
#2  0x0000555555680a8e in RawFileClass::Seek (this=0x7fffffffd790, pos=0, 
    dir=1) at /home/giulianob/bagunca/Vanilla-Conquer/common/rawfile.cpp:638
#3  0x0000555555669ee0 in BufferIOFileClass::Read (this=0x7fffffffd790, 
    buffer=0x555556271130, size=768)
    at /home/giulianob/bagunca/Vanilla-Conquer/common/bfiofile.cpp:715
#4  0x000055555566af1a in CCFileClass::Read (this=0x7fffffffd790, 
    buffer=0x555556271130, size=768)
    at /home/giulianob/bagunca/Vanilla-Conquer/common/ccfile.cpp:202
#5  0x00005555555ebf53 in Init_Game ()
    at /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/init.cpp:273
#6  0x00005555555a542c in Main_Game (argc=1, argv=0x7fffffffe088)
    at /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/conquer.cpp:135
#7  0x000055555563c0ff in main (argc=1, argv=0x7fffffffe088)
    at /home/giulianob/bagunca/Vanilla-Conquer/tiberiandawn/startup.cpp:454
```
The reason for this is because fopen is returning NULL for some files, and error is not being properly checked.
After patching rawfile.cpp to show me the file it tried to open but failed, it shows me:
```
Could not read/write to file: temperat.pal
Could not read/write to file: temperat.pal
```

However, C&C95.exe binary runs fine in the same directory under Wine.

### PR DESCRIPTION
**Pull request by giulianobelinassi**
I am opening a new PR because the previous was automatically closed when I was cleaning my main branch.

If a critical file (such as temperat.pal) is not found in the MIX
files, the game crashes. This commit fixes this by checking if
opened file is not NULL.

The original crash reason was because an incorrect version of Is_Open
was being called. This was fixed by explicit calling the correct
function.

We also change DBG_FATAL_COMMON to call ShowSimpleMessageBox.

Closes #486

### CODE REVIEW

**Comment by giulianobelinassi** [comment]
Looks like Windows is not finding the SDL headers and shared libraries. Is everything OK on the CI side?

**Comment by hifi** [comment]
There are a few things that need to be done.

First is you need to move all SDL2 related code into `debugstring.cpp`. Wrap all that within `#ifdef SDL2_BUILD`. This keeps the header interface generic and whoever includes it and calls the functions don't need to know that SDL2 is being invoked even though it's implied in SDL2 builds.

Second `debugstring.cpp` needs to be moved within `common/CMakeLists.txt` to `COMMONR_SRC` and `COMMONV_SRC` so that the SDL2 build flag is being set properly and they get the different behavior. Remaster dlls don't link with SDL2 but they also need the debug print code.

Also I'm still not fully convinced the fix is correct. Glancing over once more it seems `BufferIOFileClass::Open` unconditionally returns true even if `RawFileClass::Open` fails but that's only one part of the mess that's happening. For this crash to happen the overridden `Is_Open()` needs to return true in `RawFileClass::Raw_Seek` when the raw file is not in fact open.

Though if the fix is indeed correct we need to make sure RawFileClass only calls its own members all the time and not just at a few places as that makes it inconsistent. Even the original code does have some of the calls prefixed with RawFileClass but not all.

**Comment by giulianobelinassi** [comment]
Done. Wrapped the SDL_SimpleMessageBox in a function inside debugmessage.h, and also implemented messagebox using winuser.h MessageBoxA, if available. However, I have not tested this Windows implementation and I only relied on the CI testsuite for that.

As for if the being correct, it solves the testcase which opened this PR. I did explicitly called the class own members, but I am still not sure if I covered everything. 

Do you want me to squash all these commits into a single one and open a new PR?

File: common/CMakeLists.txt [hunk 1]
```diff
@@ -182,4 +184,4 @@ if(BUILD_VANILLATD OR BUILD_VANILLARA)
     if(SDL2)
         target_compile_definitions(commonv PUBLIC SDL2_BUILD)
     endif()
-endif()
```
Comment (hifi) [inline common/CMakeLists.txt:185]: Just a slight annoyance that your editor forced a line ending. You can configure editors like Vim not to do that.

File: common/debugstring.cpp [hunk 2]
```diff
@@ -75,3 +78,21 @@ void Debug_String_File(const char* file)
...
+
+int FatalErrorMessageBox(const char* fmt, ...)
+{
+    char buffer[256];
+    va_list args;
+    va_start(args, fmt);
+    vsnprintf(buffer, 256, fmt, args);
+    va_end(args);
+
+    /* Prefer WinAPI MessageBoxA on Windows.  */
```
Comment (hifi) [inline common/debugstring.cpp:90]: Why, though? There's really no need to do that.

File: common/debugstring.cpp [hunk 3]
```diff
@@ -75,3 +78,21 @@ void Debug_String_File(const char* file)
 
     DebugState.File = fopen(file, "w");
 }
+
+int FatalErrorMessageBox(const char* fmt, ...)
+{
+    char buffer[256];
+    va_list args;
+    va_start(args, fmt);
+    vsnprintf(buffer, 256, fmt, args);
```
Comment (hifi) [inline common/debugstring.cpp:87]: Use `sizeof`, never assume the size.

File: common/debugstring.h [hunk 4]
```diff
@@ -67,14 +69,21 @@ void Debug_String_File(const char* file);
 #define DBG_ERROR(x, ...) ((void)0)
 #endif
 
+#define _DBG_FATAL_COMMON(x, ...)                                                                                      \
```
Comment (hifi) [inline common/debugstring.h:72]: I don't really see the point of this, just add the call to FatalErrorMessageBox directly to DBG_FATAL after calling Debug_String_Log? You also can't assume you can print to stdout directly, use the log function.

File: common/rawfile.cpp [hunk 5]
```diff
@@ -89,10 +90,14 @@
  *=============================================================================================*/
 void RawFileClass::Error(int error, int canretry, char const* filename)
 {
-#ifndef _WIN32
-    perror(filename);
-#endif
-    exit(1);
+    char buffer[64];
+    const char* errorstr = strerror(error);
```
Comment (hifi) [inline common/rawfile.cpp:94]: No reason to use local variables, just pass them as arguments.

File: common/rawfile.cpp [hunk 6]
```diff
@@ -89,10 +90,14 @@
...
+    const char* fmtstring = "Error on file %s: %s\n";
+
+    printf(fmtstring, filename, errorstr);
```
Comment (hifi) [inline common/rawfile.cpp:97]: Use DBG_ERROR instead of printf.

File: common/rawfile.cpp [hunk 7]
```diff
@@ -268,13 +274,6 @@ int RawFileClass::Open(int rights)
             break;
         }
 
-        /*
```
Comment (hifi) [inline common/rawfile.cpp:271]: I don't think this needs to be moved when the error handling works as intended.

**Review by hifi — CHANGES_REQUESTED** [review]
I accidentally used the single comment button a few times. Only minor comments anymore.

You can squash the commits within a PR by force pushing to your branch.

I still want an opinion from @OmniBlade what we should do with the file classes. The functions are virtual for a reason but it all falls apart and we need to be careful not to botch it further.

Anyway good work, sorry for being unresponsive for a while.

**Comment by hifi** [comment]
Will you continue this PR?

**Comment by giulianobelinassi** [comment]
> Will you continue this PR?

Not for a while. If someone wants to continue it, fell free.

**Comment by OmniBlade** [comment]
It seems that the issue with Error being called was added during the refactor for case insensitivity. The original code has the function as an empty place holder that did not exit on error. Perhaps worth removing the fatal aspect of the function and change it to just log the error instead?

**Comment by OmniBlade** [comment]
Supercded by #735

**Pull request closed without merging**